lang/wasm_instance_cache.cc

```diff
@@ -0,0 +1,209 @@
+    auto linker = wasmtime::Linker(ctx.engine_ptr->get());
+    auto wasi_def = linker.define_wasi();
+    if (!wasi_def) {
+        throw wasm::exception(format("Setting up wasi failed: {}", wasi_def.err().message()));
+    }
+    auto cfg = wasmtime::WasiConfig();
+    auto set_cfg = store.context().set_wasi(std::move(cfg));
+    if (!set_cfg) {
+        throw wasm::exception(format("Setting up wasi failed: {}", set_cfg.err().message()));
+    }
+    auto instance_res = linker.instantiate(store, *ctx.module);
+    if (!instance_res) {
+        throw wasm::exception(format("Creating a wasm runtime instance failed: {}", instance_res.err().message()));
+    }
+    auto instance = instance_res.unwrap();
+    auto function_obj = instance.get(store, ctx.function_name);
+    if (!function_obj) {
+        throw wasm::exception(format("Function {} was not found in given wasm source code", ctx.function_name));
+    }
+    wasmtime::Func* func = std::get_if<wasmtime::Func>(&*function_obj);
+    if (!func) {
+        throw wasm::exception(format("Exported object {} is not a function", ctx.function_name));
+    }
+    auto memory_export = instance.get(store, "memory");
+    if (!memory_export) {
+        throw wasm::exception("memory export not found - please export `memory` in the wasm module");
+    }
+    auto memory = std::get<wasmtime::Memory>(*memory_export);
+
+    return wasm::wasm_instance{.store=std::move(store), .instance=std::move(instance), .func=std::move(*func), .memory=std::move(memory)};
+}
+
+// lru must not be empty, and its elements must refer to entries in _cache
+void instance_cache::evict_lru() noexcept {
+    auto& entry = _lru.front();
+    _total_size -= entry.instance_size;
+    _lru.pop_front();
+    auto it = _cache.find(entry.key);
+    _cache.erase(it);
+}
+
+void instance_cache::on_timer() noexcept {
+    auto now = seastar::lowres_clock::now();
+    while (!_lru.empty() && _lru.front().timestamp + _timer_period < now) {
+        evict_lru();
+    }
+}
+
+seastar::future<instance_cache::value_type> instance_cache::get(const key_type& key, wasm::context& ctx) {
+    cache_key_type key_pair{seastar::current_scheduling_group(), key};
+    auto it = _cache.find(key_pair);
+    if (it == _cache.end()) {
+        ++shard_stats().cache_misses;
+        _cache.try_emplace(key_pair);
+    } else {
+        ++shard_stats().cache_hits;
+    }
+    auto& i = _cache.at(key_pair);
+    if (!i.instance) {
+        auto undo = seastar::defer([this, &key_pair] {
+            _cache.erase(key_pair);
+        });
+        i.instance = seastar::make_lw_shared<wasm_instance>(load(ctx));
+        _lru.push_back({lookup_cache_key_type(key_pair), seastar::lowres_clock::now(), i.instance->memory.size(i.instance->store)});
+        undo.cancel();
+        i.it = --_lru.end();
+        _total_size += i.it->instance_size;
+    }
+    // cache entry for key_pair shouldn't be removed until we 'recycle' the instance
+    if (!i.mutex.try_lock()) {
```

Comment:
looks like shared_mutex needs a similar API to https://github.com/scylladb/seastar/pull/1126 , where you can just check if it's free to take without actually taking it yet. Can be done as a follow-up I guess